bot/src/bot.rs

```diff
@@ -0,0 +1,246 @@
+            fetch_contracts_cooldown,
+            last_fetch: SystemTime::UNIX_EPOCH,
+            contract_instances_to_ac: Default::default(),
+            autocompound_cooldown,
+        }
+    }
+
+    // Fetches contracts if fetch cooldown passed
+    pub fn fetch_contracts(&mut self) -> anyhow::Result<()> {
+        // Don't fetch if not ready
+        let ready_time = self.last_fetch + self.fetch_contracts_cooldown;
+        if SystemTime::now() < ready_time {
+            return Ok(());
+        }
+
+        let daemon = &self.daemon;
+        let mut contract_instances_to_autocompound: HashSet<(String, Addr)> = HashSet::new();
+
+        let saving_modules = self.abstract_client.version_control().module_list(
+            Some(ModuleFilter {
+                namespace: Some(self.module_info.namespace.to_string()),
+                name: Some(self.module_info.name.clone()),
+                version: None,
+                status: Some(ModuleStatus::Registered),
+            }),
+            None,
+            None,
+        )?;
+
+        for app_info in saving_modules.modules {
+            let code_id = app_info.module.reference.unwrap_app()?;
+            let mut contract_addrs = daemon
+                .rt_handle
+                .block_on(utils::fetch_instances(daemon.channel(), code_id))?;
+
+            // Only keep the contract addresses that have the required permissions
+            contract_addrs.retain(|address| {
+                utils::has_authz_permission(&self.abstract_client, address)
+                    // Don't include if queries fail.
+                    .unwrap_or_default()
+            });
+
+            // Add all the entries to the `contract_instances_to_check`
+            contract_instances_to_autocompound.extend(
+                contract_addrs
+                    .into_iter()
+                    .map(|addr| (app_info.module.info.id(), Addr::unchecked(addr))),
+            );
+        }
+
+        self.contract_instances_to_ac = contract_instances_to_autocompound;
+        Ok(())
+    }
+
+    // Autocompound all saved instances and wait for cooldown duration
+    pub fn autocompound(&self) {
+        for (id, addr) in self.contract_instances_to_ac.iter() {
+            let result = autocompound_instance(&self.daemon, (id, addr));
+            if let Err(err) = result {
+                log!(
+                    Level::Error,
+                    "error ocurred for {addr} savings-app: {err:?}"
+                );
+            }
+        }
+        // Wait for autocompound duration
+        std::thread::sleep(self.autocompound_cooldown);
+    }
+}
+
+fn autocompound_instance(daemon: &Daemon, instance: (&str, &Addr)) -> anyhow::Result<()> {
+    let (id, address) = instance;
+    let app = AppInterface::new(id, daemon.clone());
+    app.set_address(address);
+    use savings_app::AppQueryMsgFns;
+    let resp: AvailableRewardsResponse = app.available_rewards()?;
+
+    // TODO: ensure rewards > tx fee
```

Comment:
Yeah I think we should try to check this. Otherwise wallet can be easily drained

Reply:
I think tx simulation would be harder and not as accurate as using query from #8 + knowing gas consumption of this operation, especially since we are the ones who distribute this app

Reply:
We can set a limit but simulations are way more accurate. We can just use the default that our FE will set and use that as the lower bound as well.